test: checks that bitcoin-cli autocomplete is in sync

pierreN opened the pull request:
Adds a functional test that parses available RPC commands, generates the associated bitcoin-cli autocomplete file and checks that the current autocomplete file matches.

An outdated autocomplete file can be updated via the `--overwrite` test parameter.

The first commit syncs the auto-completion file with the current master. 
The second commits adds the test. It just changes the header of the autocomplete file ~~and moves it to `test/functional/data/completion/bitcoin-cli.bash-completion` (plus for ease of use and backward compatibility a symlink from `contrib/bitcoin-cli.bash-completion` to the moved file is created)~~.

This PR aims to fix: https://github.com/bitcoin/bitcoin/issues/17289

MarcoFalke commented on contrib/bitcoin-cli.bash-completion:
```diff
@@ -1,141 +0,0 @@
-# bash programmable completion for bitcoin-cli(1)
```
Why?

I'd argue the file should stay in the old directory. All the functional test needs is read-access to the file. You can get the file path in the test with:

```py
os.path.join(config["environment"]["SRCDIR"], 'contrib', ...

pierreN replied:
Thanks, I was missing the `config["environment"]` variable. Branch updated.

maflcko reviewed (approved):
Concept ACK

pierreN commented:
Hah it seems that using `SRCDIR` fails CI builds:  https://travis-ci.org/github/pierreN/bitcoin/jobs/674154306#L3417

A decent way to fix this is to add the bash auto-completion file to `Makefile.am`'s `EXTRA_DIST`. I updated the PR to reflect that.

If you think this is too intrusive another more hackish way would be to just look for the file in the parent directories: https://github.com/pierreN/bitcoin/blob/feature-auto-cli-2/test/functional/tool_cli_completion.py#L260

MarcoFalke commented on test/functional/tool_cli_completion.py:
```diff
@@ -0,0 +1,290 @@
+        """ Complete the arguments of option via the RPC help command. """
+
+        # if we can't find an argument list in the help text, then it has no option to add
+        regexp_res = ARG_REGEXP.match(self.nodes[0].help(option.command))
```
```suggestion
        res = self.nodes[0].format(command=option.command, output=args_cli)
```

Instead of using regex to parse the pseudo doc-json, a hidden RPC could be added that spits out the format you want.

I think all you need to do is register a formatter in RPCMan::Check

pierreN replied:
Yes, good idea. I have updated the branch to do that. Please have a look at the new code, thanks.

Note that some non-hidden RPC call didn't use RPCHelpMan::Check so I added a commit to fix that first (otherwise they wouldn't appear in the bash cli completion file).

pierreN replied:
I hadn't notice #18531 when writing 412bb2f - I'll rebase this PR after #18531 is merged.

MarcoFalke commented on Makefile.am:
```diff
@@ -45,7 +45,8 @@ OSX_QT_TRANSLATIONS = da,de,es,hu,ru,uk,zh_CN,zh_TW
 	       $(top_srcdir)/contrib/linearize/linearize-data.py \
-	       $(top_srcdir)/contrib/linearize/linearize-hashes.py
+	       $(top_srcdir)/contrib/linearize/linearize-hashes.py \
+	       $(top_srcdir)/contrib/bitcoin-cli.bash-completion
```
Looks good this way, if it makes out-of-tree builds happy. :+1:

Commit:
rpc: make all unhidden actors use RPCHelpMan::Check

Commit:
rpc: add format command with support for args_cli

- add format command to get infos about commands via a particular format
- add output format args_cli to get arguments type info of shown commands
- refactor RPCArg::ToTypeString to be used accross multiple output formats
- overload CRPCTable::execute to call != methods than in their request

Commit:
contrib: sync bitcoin-cli bash autocompletion

Commit:
test: keeps bitcoin-cli autocomplete in sync

Add a functional test which parses available RPC commands, generates
the associated bitcoin-cli autocomplete file and checks that the current
autocomplete file matches.
An outdated autocomplete file can be updated via the --overwrite test
parameter.

pierreN commented:
Updated branch to partially get rid of conflict with #17356 (some conflict might be left if git is not clever enough).

Added a comment for #12674 : https://github.com/bitcoin/bitcoin/pull/12674/files#r408113386

adamjonas commented:
@pierreN is this something you plan on continuing to work on?

adamjonas commented:
@pierreN ping for rebase.

pierreN commented:
Sorry for the few months delay.
I have a bit more time now and will try to follow through with this PR.

niVelion commented on contrib/bitcoin-cli.bash-completion:
```diff
@@ -1,5 +1,10 @@
+# If you want to modify this file, modify test/functional/tool_cli_completion.py and re-autogenerate
+# this file via the --overwrite test flag.
+
+# Copyright (c) 2012-2020 The Bitcoin Core developers
```
Happy New Year :clinking_glasses: 

```suggestion
# Copyright (c) 2012-2022 The Bitcoin Core developers
```

niVelion commented:
ACK, would be great to see this merged.

I can see in https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#squashing-commits that the next task is primarily to write a new commit message, the one that'll make it into the `master` branch to represent this work.

Here's the starting point:

```
# This is a combination of 4 commits.
# This is the 1st commit message:

rpc: make all unhidden actors use RPCHelpMan::Check

# This is the commit message #2:

rpc: add format command with support for args_cli

- add format command to get infos about commands via a particular format
- add output format args_cli to get arguments type info of shown commands
- refactor RPCArg::ToTypeString to be used accross multiple output formats
- overload CRPCTable::execute to call != methods than in their request

# This is the commit message #3:

contrib: sync bitcoin-cli bash autocompletion

# This is the commit message #4:

test: keeps bitcoin-cli autocomplete in sync

Add a functional test which parses available RPC commands, generates
the associated bitcoin-cli autocomplete file and checks that the current
autocomplete file matches.
An outdated autocomplete file can be updated via the --overwrite test
parameter.
```

fanquake commented:
#25243 has picked this up.
